src/passes/ifStatementTempVarPostpender.ts

```diff
@@ -0,0 +1,200 @@
+import { TranspileFailedError } from '../utils/errors';
+import { CairoFunctionDefinition } from '../ast/cairoNodes';
+import { createBlock, createCairoTempVar } from '../utils/nodeTemplates';
+import { FunctionStubKind } from '../ast/cairoNodes/cairoFunctionDefinition';
+import { hasPathWithoutReturn } from '../export';
+
+/*
+ * This pass does a live variable analysis of the code while simultaneously
+ * postpending tempvars of the live variables at the end of each if branch and
+ * after each IfStatement in order to prevent against revoked references.
+ *
+ * A variable is 'live' at a line if there is a reference to the variable after
+ * the line in the control flow for which there is no matching declaration.
+ *
+ * This following:
+ *
+ ********************
+ * // y is not live
+ * let y = 0
+ *   // y is live
+ *   if (x == 0) {
+ *   // y is live
+ * } else {
+ *   // y is not live
+ *   let y = 4;
+ *   // y is live
+ * }
+ * g(y) // reference to y
+ ********************
+ *
+ * Becomes:
+ *
+ ********************
+ * // y is not live
+ * let y = 0
+ * // y is live
+ * if (x == 0) {
+ *   // y is live
+ *   tempvar y = y;
+ * } else {
+ *   // y is not live
+ *   let y = 4;
+ *   // y is live
+ *   tempvar y = y;
+ * }
+ * tempvar y = y;
+ * g(y) // reference to y
+ ********************
+ */
+
+export class IfStatementTempVarPostpender extends ASTMapper {
+  private liveVars: Set<string> = new Set();
+
+  // Function to add passes that should have been run before this pass
+  addInitialPassPrerequisites(): void {
+    const passKeys: Set<string> = new Set<string>([
+      'An', // Most of the functions this pass removes are generated by the IfFunctionaliser
+    ]);
+    passKeys.forEach((key) => this.addPassPrerequisite(key));
+  }
+
+  visitCairoFunctionDefinition(node: CairoFunctionDefinition, ast: AST): void {
+    this.liveVars = new Set<string>();
+    [...node.implicits].forEach((implicit) => this.liveVars.add(implicit));
+    // node.vReturnParameters.vParameters.forEach((p) => this.liveVars.add(p.name));
+    // analyse the block
+    if (node.vBody) {
+      this.dispatchVisit(node.vBody, ast);
+    }
+    // remove the parameters and implicits
+    [...node.implicits].forEach((implicit) => {
+      if (!this.liveVars.delete(implicit)) {
+        throw new TranspileFailedError(
+          `Implicit variable dead. ${implicit} in function ${node.name}.`,
+        );
+      }
+    });
+    node.vParameters.vParameters.forEach((p) => {
+      if (!this.liveVars.delete(p.name) && node.functionStubKind === FunctionStubKind.None) {
+        node.functionStubKind;
+        // console.log(`Dead parameter found ${p.name} in function ${node.name}`);
+      }
+    });
+    this.liveVars = new Set<string>();
+  }
+
+  visitIdentifier(node: Identifier, _ast: AST): void {
+    if (
+      node.vReferencedDeclaration &&
+      node.vReferencedDeclaration.getClosestParentByType(CairoFunctionDefinition)
```

Comment:
I assume this declaration should be in the same function definition as the identifier, but this condition does not ensure that. I think I would like some kind of assertion here to ensure it.